ELClient/ELClientWebServer.cpp

```diff
@@ -0,0 +1,407 @@
+}
+
+void ELClientWebServer::unregisterHandler(const __FlashStringHelper * URL)
+{
+  String s = URL;
+  unregisterHandler(s);
+}
+
+// unregisters a previously registered handler
+void ELClientWebServer::unregisterHandler(const String &URL)
+{
+  struct _Handler *prev = 0;
+  struct _Handler *hnd = handlers;
+  while( hnd != 0 )
+  {
+    if( hnd->URL == URL )
+    {
+      if( prev == 0 )
+        handlers = hnd->next;
+      else
+        prev->next = hnd->next;
+      
+      delete hnd;
+    }
+    prev = hnd;
+    hnd = hnd->next;
+  }
+}
+
+void ELClientWebServer::registerCallback()
+{
+  // WebServer doesn't send messages to MCU only if asked
+  // register here to the web callback
+  // periodic reregistration is required in case of ESP8266 reset
+  _elc->Request(CMD_CB_ADD, 100, 1);
+  _elc->Request(F("webCb"), 5);
+  _elc->Request();
+}
+
+void ELClientWebServer::processPacket(ELClientPacket *packet)
+{
+  ELClientResponse response(packet);
+
+  uint16_t shrt;
+  response.popArg(&shrt, 2);
+  RequestReason reason = (RequestReason)shrt; // request reason
+
+  response.popArg(remote_ip, 4);    // remote IP address
+  response.popArg(&remote_port, 2); // remote port
+  
+  char * url;
+  int urlLen = response.popArgPtr(&url);
+  
+  WebServerCallback callback = 0;
+  
+  struct _Handler *hnd = handlers;
+  while( hnd != 0 )
+  {
+    if( hnd->URL.length() == urlLen && memcmp( url, hnd->URL.begin(), urlLen ) == 0 )
+    {
+      callback = hnd->callback;
+      break;
+    }
+    hnd = hnd->next;
+  }
+
+  if( hnd == 0 ) // no handler found for the URL
+  {
+    _elc->_debug->print(F("Handler not found for URL:"));
+    
+    for(int i=0; i < urlLen; i++)
+      _elc->_debug->print( url[i] );
+    _elc->_debug->println();
+    return;
+  }
+  
+  switch(reason)
+  {
+    case WS_BUTTON: // invoked when a button pressed
+      {
+        char * idPtr;
+        int idLen = response.popArgPtr(&idPtr);
+  
+        char id[idLen+1];
+        memcpy(id, idPtr, idLen);
+        id[idLen] = 0;
+
+        callback(BUTTON_PRESS, id, idLen);
+      }
+      break;
+    case WS_SUBMIT: // invoked when a form submitted
+      {
+        int cnt = 4;
+
+        while( cnt < response.argc() )
+        {
+          char * idPtr;
+          int idLen = response.popArgPtr(&idPtr);
+          int nameLen = strlen(idPtr+1);
+          int valueLen = idLen - nameLen -2;
+
+          arg_ptr = (char *)malloc(valueLen+1);
+          arg_ptr[valueLen] = 0;
+          memcpy(arg_ptr, idPtr + 2 + nameLen, valueLen);
+
+          callback(SET_FIELD, idPtr+1, nameLen);
+  
+          free(arg_ptr);
+          arg_ptr = 0;
+          cnt++;
+        }
+      }
+      return;
+    case WS_LOAD: // invoked at refresh / load
+    case WS_REFRESH:
+      break;
+    default:
+      return;
+  }
+  
+  // the response is generated here with the fields to refresh
+
+  _elc->Request(CMD_WEB_DATA, 100, 255);     // 255 arguments (can't foretell, so use maximum)
```

Comment:
Would be best to define some notion of "varargs" and use a named constant for 255 and the also in esp-link... Basically make 255 args mean "varargs" officially.